File: cpp_files/raylib/shape_survivors/src/ui_manager.cpp
#include "settings.hpp"
#include "ui_manager.hpp"
#include "utils.hpp"

UiManager::UiManager(const Player& player, const EnemyManager& enemy_manager)
    : player(player), enemy_manager(enemy_manager) {}

void UiManager::draw_player_xp_bar(void) {
    const float     margin = 50.0f;
    const Rectangle bounds = {margin, 10.0f, BASE_WIDTH - margin * 2, 30.0f};

    Utils::draw_progress_bar(
        player.xp.current, player.xp.next, bounds, 3.0f, BLACK, LIGHTGRAY, BLUE);

    // draw player level
    const float offset = 5.0f;
    DrawText(TextFormat("LV %d", player.level.current),
             static_cast<int>(BASE_WIDTH - margin * 2),
             static_cast<int>(bounds.y + bounds.height + offset),
             20,
             BLACK);
}

void UiManager::draw_player_hp_bar(void) {
    const float width  = player.rect.width + 10;
    const float height = 10.0f;
    const float x      = player.position.x - width / 2.0f;
    const float y      = player.position.y - player.rect.height / 2.0f - height * 2;

    Utils::draw_progress_bar(player.health.current,
                             player.health.max,
                             {x, y, width, height},
                             1.0f,
                             BLACK,
                             LIGHTGRAY,
                             GREEN);
}

void UiManager::draw_enemy_hp_bar(void) {
    for (const Enemy& enemy : enemy_manager.enemies) {
        const float width  = enemy.rect.width + 10;
        const float height = 10.0f;
        const float x      = enemy.position.x - width / 2.0f;
        const float y      = enemy.position.y - enemy.rect.height / 2.0 - height * 2;

        Utils::draw_progress_bar(enemy.health.current,
                                 enemy.health.max,
                                 {x, y, width, height},
                                 1.0f,
                                 BLACK,
                                 LIGHTGRAY,
                                 RED);
    }
}

File: python_files/raylib_files/shape_survivors/ui_manager.py
import pyray as p

import utils
from enemy_manager import EnemyManager
from player import Player
from settings import BASE_WIDTH


class UiManager:
    def __init__(self, player: Player, enemy_manager: EnemyManager) -> None:
        self.player: Player = player
        self.enemy_manager: EnemyManager = enemy_manager

    def draw_player_xp_bar(self) -> None:
        margin: float = 50.0
        bounds: p.Rectangle = p.Rectangle(margin, 10.0, BASE_WIDTH - margin * 2, 30.0)

        utils.draw_progress_bar(
            min_=self.player.xp.current,
            max_=self.player.xp.next,
            bounds=bounds,
            line_thickness=3.0,
            line_color=p.BLACK,
            bg_color=p.LIGHTGRAY,
            fill_color=p.BLUE,
        )

        # draw player level
        offset: float = 5.0
        p.draw_text(
            f"LV {self.player.level.current}",
            int(BASE_WIDTH - margin * 2),
            int(bounds.y + bounds.height + offset),
            20,
            p.BLACK,
        )

    def draw_player_hp_bar(self) -> None:
        width: float = self.player.rect.width + 10
        height: float = 10.0
        x: float = self.player.position.x - width / 2.0
        y: float = self.player.position.y - self.player.rect.height / 2.0 - height * 2

        utils.draw_progress_bar(
            min_=self.player.health.current,
            max_=self.player.health.max,
            bounds=p.Rectangle(x, y, width, height),
            line_thickness=1.0,
            line_color=p.BLACK,
            bg_color=p.LIGHTGRAY,
            fill_color=p.GREEN,
        )

    def draw_enemy_hp_bar(self) -> None:
        for enemy in self.enemy_manager.enemies:
            width: float = enemy.rect.width + 10
            height: float = 10.0
            x: float = enemy.position.x - width / 2.0
            y: float = enemy.position.y - enemy.rect.height / 2.0 - height * 2

            utils.draw_progress_bar(
                min_=enemy.health.current,
                max_=enemy.health.max,
                bounds=p.Rectangle(x, y, width, height),
                line_thickness=1.0,
                line_color=p.BLACK,
                bg_color=p.LIGHTGRAY,
                fill_color=p.RED,
            )
